app/models/auth_factor.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# An authentication factor used for two factor authentication.
+class AuthFactor < ApplicationRecord
+  include PaperTrailable
+
+  cattr_accessor :form_steps do
+    [:registration, :verification, :confirmation]
+  end
+  attr_accessor :form_step
+
+  before_validation :clean_fields
+
+  belongs_to :user
+
+  CHANNELS = [
+    SMS = 'sms'
+  ].freeze
+
+  CHANNELS.each do |channel|
+    scope channel, -> { where(channel: channel) }
+  end
+
+  validates :channel, presence: true, inclusion: { in: CHANNELS }
+
+  with_options if: -> { required_for_step?(:registration) } do
+    validates :name, presence: true, uniqueness: { scope: :user_id }, length: { maximum: 30 }
+    validates :phone, presence: true, format: /\A\d{10}\z/, length: { is: 10 }
+  end
+
+  private
+
+  def clean_fields
+    phone&.gsub!(/\D/, '')
+    name&.strip!
+  end
+
+  def required_for_step?(step)
+    # All fields are required if no form step is present
+    return true if form_step.nil?
+
+    # All fields from previous steps are required if the
+    # step parameter appears before or we are on the current step
+    return true if form_steps.index(step) <= form_steps.index(form_step)
+  end
```

Comment:
You're right, it won't ever return false when called with the first step like we're doing here since we're always either on or after the first step. It's more intended for conditional validations for later steps where we wouldn't want to validate before we've actually gotten to that step. I got this method from [this guide](https://jonsully.net/blog/rails-wizards-part-four/) and I think that example makes it clearer.

Even though it's not strictly necessary to have this right now, I still think it's nice for readability because by wrapping validations in this check you essentially say "these validations are on fields we get in the registration step". I don't feel strongly about this at all though so I'm happy to remove and just accomplish the same thing with a comment if you'd prefer.

Reply:
Moved some wording around in the latest revision to hopefully make this method a bit more readable.